pkg/controller/datavolume-controller.go

```diff
@@ -373,194 +383,319 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
-					}
+	srcCapacity, hasSrcCapacity := sourcePvc.Status.Capacity[corev1.ResourceStorage]
+	targetRequest, hasTargetRequest := targetStorageSpec.Resources.Requests[corev1.ResourceStorage]
+	allowExpansion := srcStorageClass.AllowVolumeExpansion != nil && *srcStorageClass.AllowVolumeExpansion
+	if !hasSrcCapacity || !hasTargetRequest || (srcCapacity.Cmp(targetRequest) < 0 && !allowExpansion) {
+		// return error so we retry the reconcile
+		return false, errors.New("source/target sizes not compatible")
+	}
 
-					// trigger transfer and next reconcile should have pvcExists == true
-					tmpPVC.Annotations[annReadyForTransfer] = "true"
-					tmpPVC.Annotations[AnnPopulatedFor] = datavolume.Name
-					if err := r.client.Update(context.TODO(), tmpPVC); err != nil {
-						return reconcile.Result{}, err
-					}
+	return true, nil
+}
 
-					return reconcile.Result{},
-						r.updateSmartCloneStatusPhase(cdiv1.NamespaceTransferInProgress, datavolume, nil)
-				} else {
-					return reconcile.Result{}, nil
-				}
+func (r *DatavolumeReconciler) selectCloneStrategy(datavolume *cdiv1.DataVolume, pvcSpec *corev1.PersistentVolumeClaimSpec) (cloneStrategy, error) {
+	doSmartClone := false
+	doCsiClone := false
+
+	isClone := false
+	if datavolume.Spec.Source.PVC != nil {
+		isClone = true
+
+		// The source and target PVCs share the same Storage Class
+		// get this the same way as for getSnapshotClassForSmartClone(datavolume, pvcSpec)
+		preferredCloneStrategy, err := r.getCloneStrategy(datavolume)
+		if err != nil {
+			return 0, err
+		}
+
+		if isClone && *preferredCloneStrategy == cdiv1.CloneStrategyCsiClone {
+			possible, err := r.isCSIClonePossible(datavolume, pvcSpec)
+			if err != nil {
+				return 0, err
 			}
 
-			if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
-				return reconcile.Result{}, nil
+			doCsiClone = possible
+		} else if isClone && *preferredCloneStrategy == cdiv1.CloneStrategySnapshot {
+			bindingMode, err := r.getStorageClassBindingMode(pvcSpec.StorageClassName)
+			if err != nil {
+				return 0, err
 			}
 
-			r.log.V(3).Info("Smart-Clone via Snapshot is available with Volume Snapshot Class",
-				"snapshotClassName", snapshotClassName)
+			snapshotClassName, err := r.getSnapshotClassForSmartClone(datavolume, pvcSpec)
+			if err != nil {
+				return 0, err
+			}
+			snapshotClassAvailable := snapshotClassName != ""
 
-			newSnapshot := newSnapshot(datavolume, pvcName, snapshotClassName)
-			if err := setAnnOwnedByDataVolume(newSnapshot, datavolume); err != nil {
-				return reconcile.Result{}, err
+			snapshotPossible, err := r.snapshotSmartClonePossible(datavolume, pvcSpec)
+			if err != nil {
+				return 0, err
 			}
 
-			nn := client.ObjectKeyFromObject(newSnapshot)
-			if err := r.client.Get(context.TODO(), nn, newSnapshot.DeepCopy()); err != nil {
-				if !k8serrors.IsNotFound(err) {
-					return reconcile.Result{}, err
-				}
+			doSmartClone = snapshotClassAvailable && snapshotPossible &&
```

Comment:
```go
if snapshotClassAvailable && snapshotPossible &&
  (!isCrossNamespaceClone(datavolume) || *bindingMode == storagev1.VolumeBindingImmediate) {
  return SmartClone, nil
}
This also means we can't do cross namespace clone on WFFC storage? Why does this not also apply to CSI clone?

Reply:
Yes this logic is unchanged, just extracted to a method `selectCloneStrategy`

And I think you are right, CSI clone has the same. Because with WFFC you would need to schedule a pod to bind a volume, and so far we do not have logic for that. I do not know if we can correctly do it. 

On the other hand I am not sure why exactly this condition
`(!isCrossNamespaceClone(datavolume) || *bindingMode == storagev1.VolumeBindingImmediate) {` 

@mhenriks can you comment on that?